Update tailwindcss `@apply` suggestion

bluwy opened the pull request:
## Changes

There has been a couple issues about tailwindcss recently:

- #5384 
- #5390

This PR clarifies that inline classes should be preferred instead of `@apply` as it does not work for classes like `dark:text-pink` and duplicates the global and Astro scoped style output.

Question: Should we mentioned that it's not possible to code-split tailwind CSS output per-page too? All pages can currently only share one stylesheet file.

## Testing

<!-- How was this change tested? -->
<!-- DON'T DELETE THIS SECTION! If no tests added, explain why. -->
N/A

## Docs

<!-- Could this affect a user’s behavior? We probably need to update docs! -->
<!-- If docs will be needed or you’re not sure, uncomment the next line: -->
<!-- /cc @withastro/maintainers-docs for feedback! -->

<!-- DON'T DELETE THIS SECTION! If no docs added, explain why.-->
<!-- https://github.com/withastro/docs -->
Updated tailwindcss README

matthewp commented:
Update looks good to me.

delucis commented on packages/integrations/tailwind/README.md:
```diff
@@ -127,7 +127,9 @@ export default {
 You can now [import your own `base.css` as a local stylesheet](https://docs.astro.build/en/guides/styling/#import-a-local-stylesheet).
 
-If you are using Vue, Svelte, or another component integration with Astro, `@apply` directives used in component `<style>`s may generate errors about your custom Tailwind class not existing and cause your builds to fail. [Instead of using `@layer` directives in a a global stylesheet](https://tailwindcss.com/docs/functions-and-directives#using-apply-with-per-component-css), define your custom styl ...
+When using the `@apply` directive in an Astro, Vue, Svelte, or another component integration's `<style>` tag, it may generate errors about your custom Tailwind class not existing and cause your builds to fail. Using nested classes like `dark:text-pink` would also not work in scoped `<style>`.
```
Do we have a clearer idea about what exactly triggers these errors? I’ve used `@apply` in `.astro` successfully, e.g. <https://github.com/withastro/astro.build/blob/a4d8460976a36e335f2b0b4289e277b190750e19/src/components/CodeHero.astro#L65-L89>

Is it only nested classes that are an issue?

bluwy replied:
If I understand correctly from previous reports, classes in `@apply` are also extracted into the main tailwind css file, so there's two copies of the class. Maybe it's not obvious in a large project where it's used in several places.

delucis replied:
Ahh, I see. Sounds like there are actually two (or three?) issues:

1. Some usage triggers “custom Tailwind class not existing” errors, breaking builds.

2. _Any_ usage can result in duplicate CSS, because Tailwind scans the `.astro` file to know which classes to include in the main bundle and is not smart enough to distinguish between a class name in frontmatter, a `class` attribute, etc. or inside `<style>`.

3. Nested classes like `dark:text-pink` are not compatible with `@apply` (is this actually the issue that causes 1?)

I think for **1** it would be amazing if we have more specific guidance than “may generate” if we know which usage is the issue (we may not of course).

For **2**, I think this can probably framed in terms of best practices like you’re doing in the second paragraph you added. Maybe we also state clearly there how you end up with duplicate CSS.

bluwy replied:
Just tested no2 and it's false. Turns out tailwind strips of classes from `@apply` so not sure why someone mentioned it doesn't 😄 Will remove that.

For no1, it seems to be added by https://github.com/withastro/astro/pull/4353 about custom classes.

For no3, it's a separate issue from no1. `dark:text-pink` classes specifically is generated as `.dark > .text-pink`, but this doesn't match any element in the Astro markup so the compiler shakes them out.

delucis reviewed (commented):
Thanks @bluwy! If I understand correctly, this text and the code sample that follow are not strictly speaking related to the `config.applyBaseStyles` option, correct? We just added these there because we were already talking about `@layer`?

I wonder if we should move this to the “Troubleshooting” section with a dedicated subheading like “Using `@apply` with components”?

bluwy commented:
> I wonder if we should move this to the “Troubleshooting” section with a dedicated subheading like “Using `@apply` with components”?

Make sense to me. I wonder why the previous docs were there too but this definitely sounds like a troubleshooting guide. Will move it there.

bluwy commented:
Updated the docs and moved the two `@apply` issues under troubleshooting. GitHub's diff isn't showing it right, but I didn't touch the examples section.

tony-sull commented:
For issue #1, is that a Tailwind issue in general, or does it have something to do with how Tailwind picks up the config file in an astro project?

For #3, I think something like `h1 { @apply dark:text-gray-50; }` should be compiled directly to the code below, with our scoped class added as usual.  Maybe there are certain edge cases where it breaks?

```css
@media (prefers-color-scheme: dark) {
 h1:where(.astro-GFDZHAND) {
  --tw-text-opacity: 1;
  color:rgb(249 250 251 / var(--tw-text-opacity))
 }
}
```

bluwy commented:
> For issue (1), is that a Tailwind issue in general, or does it have something to do with how Tailwind picks up the config file in an astro project?

It's a tailwind issue in general added in https://github.com/withastro/astro/pull/4353. The tailwind docs also outlined this caveat: https://tailwindcss.com/docs/functions-and-directives#using-apply-with-per-component-css

Yup re no3, it only happens for `darkMode: "class"`

delucis commented on packages/integrations/tailwind/README.md:
````diff
@@ -146,14 +164,18 @@ module.exports = {
 ```
 
-## Examples
+### Class-based directives like `dark:text-pink` with `@apply` does not work
````
It looks like [the Tailwind docs](https://tailwindcss.com/docs/hover-focus-and-other-states) call `dark:*`, `hover:*` etc. “modifiers”. I’ve also tweaked this heading to match the previous heading format.

```suggestion
### Class with modifiers does not work with `@apply` directives
```

bluwy replied:
The issue is that only some modifiers do not work though. `hover:` and `focus:` etc works because they generate `.class:hover`. `dark:` however generates `.dark .class` which is causing problems in scoped styles.

Maybe we should call them class-based modifiers?

delucis replied:
Ahhhh, thank you! I had not understood that nuance.

Then, yes, maybe this heading would be more accurate:

```suggestion
### Class-based modifiers do not work with `@apply` directives
```

delucis commented on packages/integrations/tailwind/README.md:
```diff
@@ -146,14 +164,18 @@ module.exports = {
-- Astro's homepage uses Tailwind. Check out its [Tailwind configuration file](https://github.com/withastro/astro.build/blob/main/tailwind.config.cjs) or an [example component](https://github.com/withastro/astro.build/blob/main/src/components/IntegrationCard.astro)
-- The [Astro Ink](https://github.com/one-aalam/astro-ink), [Sarissa Blog](https://github.com/iozcelik/SarissaBlogAstroStarter), and [Creek](https://github.com/robertguss/Astro-Theme-Creek) themes use Tailwind for styling
-- [Browse Astro Tailwind projects on GitHub](https://github.com/search?q=%22%40astrojs%2Ftailwind%22+filename%3Apackage.json&type=Code) for more examples!
+Astro `<style>` tags are scoped by default. As `dark:text-pink` compiles into `.dark .text-pink`, it does not match any markup in the Astro file and is hence removed, causing the class to not be applied. This may also happen for integrations like Vue and Svelte that supports scoped styles.
```
Reworked this slightly to move the example class down from the subheading.

```suggestion
Class names using modifiers like `dark:text-pink` or `focus:ring` are not compatible with the `@apply` directive in Astro `<style>` tags. For example, `dark:text-pink` compiles to `.dark .text-pink`, which is removed from build output because it does not match any markup in the `.astro` file. The same issue may also happen in framework components that support scoped styles like Vue and Svelte.
```

delucis replied:
Given the above, maybe this is more accurate? Are there any other good examples where this is the case? I guess the sample applies with the group classes like `group-hover:*`?

```suggestion
Certain Tailwind classes with modifiers rely on combining classes across multiple elements. For example, `dark:text-pink` compiles to `.dark .text-pink`. When this is used with the `@apply` directive in Astro `<style>` tags, the compiled styles are removed from the build output because they do not match any markup in the `.astro` file. The same issue may also happen in framework components that support scoped styles like Vue and Svelte.
```

bluwy replied:
Yep `group-hover:*` should also have the same issue. I think that's a better example too as `dark:` could be configured to use media queries. I'll swap that out after committing your suggestion

delucis reviewed (approved):
Thanks @bluwy! I think these are really good and provide clear examples of both the problem and what to do instead. I’ve suggested some wording tweaks but then should be good to go 🙌

delucis reviewed (approved):
Thanks for working through this with me @bluwy!
